t/50reverse-proxy-early-response.t

```diff
@@ -70,7 +70,7 @@ EOS
                 my $client = H1Client->new($server);
                 $client->send_headers('POST', '/', ['transfer-encoding' => 'chunked']) or die $!;
                 $client->send_data("1\r\na\r\n") or die $!;
-                sleep 0.01;
+                sleep 1;
```

Comment:
I do not understand why we need to sleep here at all. Isn't `$client->read(1000)` right below doing a blocking read of a specified duration?

Reply:
@kazuho @gfx Now in this test case, we're checking the connection is properly closed by trying `read` with `MSG_PEEK` and seeing its response and errno. The failing scenario seems to be

1. h2o sends response header
2. `$client->read(1000)` reads only that header and proceed to the next
3. While doing other stuff (e.g. $client->send_data), h2o is still sending the response body to the client (with TE: chunked)
4. `$client->is_alive` returns true because it can read response body in kernel receive buffer, though it has already sent FIN

If the above scenario can possibly happen, adding sleep is a (dirty but effective) workaround because it mostly ensures h2o sends whole response body before the client starts reading. OR, while writing this comment I've came up with an another idea that adding `read_all` function, which read all data from the socket until it gets closed, should be a more robust way to avoid freakiness and slowing test. WDYT?

Reply:
@i110 Thank you for elaborating. FWIW h2o sends header and body at once provided that the upstream sends them at once. Therefore I assume that you point is based on the understanding that the upstream sends them separately.

That said, I agree that that is a possibility.

Though then the question is why do we need to do `read(1000)` that invokes select followed by sysread? To be clear, my argument is that:
* If we are to postpone reading from the socket until we can assume that all necessary data will be available, we do not need to do a select-read. We can simply do `$client->sysread(my $output, 1000)`.
* If we are to do a select-read, then we do need to have a sleep, because the assumption of doing select-read just once is that all data would be available at once.

Reply:
@kazuho 
> he upstream sends them separately.

Yes, exactly. Our implementation of h2 upstream server send them separately. See https://github.com/h2o/h2o/blob/master/t/Util.pm#L512-L514

> why do we need to do read(1000) that invokes select followed by sysread?

H1Client has non-blocking socket, and the current semantics of `read` function is `read all available data at that time`. FWIW `1000` means timeout in ms, not the desired length of the data to be read.

Reply:
Discussed offline with @i110, agreed that we no longer need to do a blocking read with timeout (i.e.,` $client->read(1000)`. Changed that line to `$client->read(0)` in 97fd81f.